kvdb-rocksdb/benches/bench_read_perf.rs

```diff
@@ -0,0 +1,205 @@
+/// points in the key set.
+fn populate(db: &Database) -> io::Result<Vec<H256>> {
+	let mut needles = Vec::with_capacity(NEEDLES);
+	let mut batch = db.transaction();
+	for i in 0..NEEDLES * NEEDLES_TO_HAYSTACK_RATIO {
+		let key = H256::random();
+		if i % NEEDLES_TO_HAYSTACK_RATIO == 0 {
+			needles.push(key.clone());
+			if i % 100_000 == 0 && i > 0 {
+				println!("[populate] {} keys", i);
+			}
+		}
+		// In ethereum keys are mostly 32 bytes and payloads ~140bytes.
+		batch.put(Some(0), &key.as_bytes(), &n_random_bytes(140));
+	}
+	db.write(batch)?;
+	// Clear the overlay
+	db.flush()?;
+	Ok(needles)
+}
+
+fn get(c: &mut Criterion) {
+	let db = open_db();
+	let needles = populate(&db).expect("rocksdb works");
+
+	let mut total_iterations = 0;
+	let mut total_allocs = 0;
+
+	c.bench_function("get key", |b| {
+		b.iter_custom(|iterations| {
+			total_iterations += iterations;
+			let mut elapsed = Duration::new(0, 0);
+			let (alloc_stats, _) = count_alloc(|| {
+				let start = Instant::now();
+				for _ in 0..iterations {
+					// This has no measurable impact on performance (~30ns)
+					let needle = needles.choose(&mut rand::thread_rng()).expect("needles is not empty");
+					black_box(db.get(Some(0), needle.as_bytes()).unwrap());
+				}
+				elapsed = start.elapsed();
+			});
+			total_allocs += alloc_stats.0;
+			elapsed
+		});
+	});
+	if total_iterations > 0 {
+		println!(
+			"[get key] total: iterations={}, allocations={}; allocations per iter={:.2}\n",
+			total_iterations,
+			total_allocs,
+			total_allocs as f64 / total_iterations as f64
+		);
+	}
+
+	total_iterations = 0;
+	total_allocs = 0;
+	c.bench_function("get key by prefix", |b| {
+		b.iter_custom(|iterations| {
+			total_iterations += iterations;
+			let mut elapsed = Duration::new(0, 0);
+			let (alloc_stats, _) = count_alloc(|| {
+				let start = Instant::now();
+				for _ in 0..iterations {
+					// This has no measurable impact on performance (~30ns)
+					let needle = needles.choose(&mut rand::thread_rng()).expect("needles is not empty");
+					black_box(db.get_by_prefix(Some(0), &needle.as_bytes()[..8]).unwrap());
+				}
+				elapsed = start.elapsed();
+			});
+			total_allocs += alloc_stats.0;
+			elapsed
+		});
+	});
+	if total_iterations > 0 {
+		println!(
+			"[get key by prefix] total: iterations={}, allocations={}; allocations per iter={:.2}\n",
+			total_iterations,
+			total_allocs,
+			total_allocs as f64 / total_iterations as f64
+		);
+	}
+}
+
+fn iter(c: &mut Criterion) {
+	let db = open_db();
+	let mut total_iterations = 0;
+	let mut total_allocs = 0;
+
+	c.bench_function("iterate over 1k keys", |b| {
+		b.iter_custom(|iterations| {
+			total_iterations += iterations;
+			let mut elapsed = Duration::new(0, 0);
+			let (alloc_stats, _) = count_alloc(|| {
+				let start = Instant::now();
+				for _ in 0..iterations {
+					black_box(db.iter(Some(0)).take(1000).collect::<Vec<_>>());
```

Comment:
I find it strange that "iterate over 1k keys" and "single key from iterator" have the same allocations per iter, the former allocates a vec here, any insights? Also we should note it doesn't count RocksDB allocations.

Reply:
On line 176 when printing the allocations I divide by `1000`, so it's easy to compare the nr allocations between the single/iterator cases. Maybe I should simply remove that and let the code reader check for themselves?

The module level docs say `//! Allocations (on the Rust side) are counted and printed.` but you're right: it definitely bears repeating.

Reply:
Removed the division by 1000 in the `println` below.